libevmasm/Inliner.cpp

```diff
@@ -0,0 +1,178 @@
+		if (item.type() == PushTag)
+			numPushTags[item.data()]++;
+
+		// We can only inline blocks with straight control flow that end in a jump.
+		// Using breaksCSEAnalysisBlock will hopefully allow the return jump to be optimized after inlining.
+		if (lastTag && SemanticInformation::breaksCSEAnalysisBlock(item, false))
+		{
+			if (item == Instruction::JUMP)
+				inlinableBlockItems[_items[*lastTag].data()] = _items | ranges::views::slice(*lastTag + 1, index + 1);
+			lastTag.reset();
+		}
+
+		if (item.type() == Tag)
+			lastTag = index;
+	}
+
+	// Filter candidates for general inlinability and store the number of PushTag's alongside the assembly items.
+	map<u256, InlinableBlock> result;
+	for (auto&& [tag, items]: inlinableBlockItems)
+		if (uint64_t const* numPushes = util::valueOrNullptr(numPushTags, tag))
+		{
+			InlinableBlock block{items, *numPushes};
+			if (isInlineCandidate(tag, block))
+				result.emplace(std::make_pair(tag, InlinableBlock{items, *numPushes}));
+		}
+	return result;
+}
+
+namespace
+{
+optional<AssemblyItem::JumpType> determineJumpType(AssemblyItem::JumpType _intoBlock, AssemblyItem::JumpType _outOfBlock)
+{
+	auto jumpTypeToInt = [](AssemblyItem::JumpType _jumpType) -> int {
+		switch (_jumpType)
+		{
+			case AssemblyItem::JumpType::IntoFunction:
+				return +1;
+			case AssemblyItem::JumpType::OutOfFunction:
+				return -1;
+			case AssemblyItem::JumpType::Ordinary:
+				return 0;
+		}
+		assertThrow(false, OptimizerException, "");
+	};
+	switch (jumpTypeToInt(_intoBlock) + jumpTypeToInt(_outOfBlock))
+	{
+		case 0:
+			return AssemblyItem::JumpType::Ordinary;
+		case 1:
+			return AssemblyItem::JumpType::IntoFunction;
+		case -1:
+			return AssemblyItem::JumpType::OutOfFunction;
+		default:
+			return nullopt;
+	}
+}
+}
+
+optional<AssemblyItem> Inliner::shouldInline(u256 const&, AssemblyItem const& _jump, InlinableBlock const& _block) const
+{
+	// Determine the exit jump to be used, if the block is inlined.
+	AssemblyItem exitJump = _block.items.back();
+	if (auto exitJumpType = determineJumpType(_jump.getJumpType(), exitJump.getJumpType()))
+		exitJump.setJumpType(*exitJumpType);
+	else
+		return nullopt;
+
+	// Always try to inline if there is at most one call to the block.
+	if (_block.pushTagCount == 1)
+		return exitJump;
+
+	// Always inline small blocks.
+	if (static_cast<size_t>(_block.items.size()) <= m_inlineMaxOpcodes)
+		return exitJump;
+
+	return nullopt;
+}
+
+
+void Inliner::optimise()
+{
+	std::map<u256, InlinableBlock> inlinableBlocks = determineInlinableBlocks(m_items);
+
+	if (inlinableBlocks.empty())
+		return;
+
+	AssemblyItems newItems;
+	for (auto it = m_items.begin(); it != m_items.end(); ++it)
+	{
+		AssemblyItem const& item = *it;
+		if (next(it) != m_items.end())
+		{
+			AssemblyItem const& nextItem = *next(it);
+			if (item.type() == PushTag && nextItem == Instruction::JUMP)
+				if (auto *inlinableBlock = util::valueOrNullptr(inlinableBlocks, item.data()))
```

Comment:
```suggestion
				if (auto inlinableBlock = util::valueOrNullptr(inlinableBlocks, item.data()))
```
Isn't it redundant? Fine with both situations. If you keep it, then `auto*` followed by whitespace, right? If you remove it, then similarly for line 148.

Reply:
It's redundant, but in the constant fight about overusing ``auto`` or not it often worked as a compromise ;-).